src/services.rs

```diff
@@ -0,0 +1,757 @@
+                        e
+                })?;
+            }
+            Event::ManifestUnpublished(mu) => {
+                let name = mu.name.clone();
+                self.handle_manifest_unpublished(mu).map_err(|e| {
+                    error!(lattice_id = %self.lattice_id, manifest = name, "Error handling manifest unpublished event: {}", e);
+                    e
+                })?;
+            }
+            _ => {}
+        }
+        Ok(())
+    }
+
+    /// Handles a manifest published event.
+    fn handle_manifest_published(&self, mp: ManifestPublished) -> Result<()> {
+        debug!(manifest=?mp, "Handling manifest published event");
+        let manifest = mp.manifest;
+        if let Some(httpserver_service) = http_server_component(&manifest) {
+            if let Some(address) = find_address(&manifest, httpserver_service.name.as_str()) {
+                debug!(address = address, "Found address");
+                if let Ok(addr) = address.parse::<SocketAddr>() {
+                    debug!("Upserting service for manifest: {}", manifest.metadata.name);
+                    self.tx
+                        .try_send(WatcherCommand::UpsertService(ServiceParams {
+                            name: manifest.metadata.name.clone(),
+                            lattice_id: self.lattice_id.clone(),
+                            port: addr.port(),
+                            namespaces: self.namespaces.clone(),
+                            host_labels: httpserver_service.labels,
+                        }))
+                        .map_err(|e| anyhow::anyhow!("Error sending command to watcher: {}", e))?;
+                } else {
+                    error!(address = address, "Invalid address in manifest");
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Handles a manifest unpublished event.
+    fn handle_manifest_unpublished(&self, mu: ManifestUnpublished) -> Result<()> {
+        self.tx
+            .try_send(WatcherCommand::RemoveService {
+                name: mu.name,
+                namespaces: self.namespaces.clone(),
+            })
+            .map_err(|e| anyhow::anyhow!("Error sending command to watcher: {}", e))?;
+        Ok(())
+    }
+}
+
+/// Waits for commands to update or remove services based on manifest deploy/undeploy events in
+/// underlying lattices.
+/// Each lattice is managed by a [Watcher] which listens for events relayed by a NATS consumer and
+/// issues commands to create or update services in the cluster.
+pub struct ServiceWatcher {
+    watchers: Arc<RwLock<HashMap<String, Watcher>>>,
+    sender: mpsc::Sender<WatcherCommand>,
+}
+
+impl ServiceWatcher {
+    /// Creates a new service watcher.
+    pub fn new(k8s_client: KubeClient) -> Self {
+        // TODO: Should this be unbounded or have a larger bound?
+        let (tx, mut rx) = mpsc::channel::<WatcherCommand>(1000);
```

Comment:
@thomastaylor312 you know Tokio internals better than I do. Is this reasonable or should it actually be unbounded? I wouldn't think it would get backed up very often if at all as long as we're not failing to ack messages